Add pricers to calculate volume prices

tvdeyen opened the pull request:
Instead of monkey patching the variant model, we leverage the new pricers from Solidus 1.3

Commit:
Add a volume_pricing_options class

For the upcoming volume pricer we need a pricing options object that has quantity and user accessors.

Commit:
Add a volume pricer

Instead of monkey patching the variant we introduce a volume pricer class, that calculates the prices.

Commit:
Remove volume_price from spree_variant_decorator

Use the pricer to display volume prices in the frontend.

Also reorganizes remaining variant specs. These specs where actually testing earning amounts not volume prices and run pretty slow.

Commit:
Move join_volume_prices from Variant to VolumePrice

Instead of monkey patching the variant class we use the VolumePrice model for finding volume prices by a variant and optional user.

Also reduces a lot of duplication in the code.

Commit:
Move volume price earning into pricer

Commit:
Add a price display class

This class is responsible for displaying volume prices and earnings and is used by the spree base helper decorator

Commit:
Move earning percent method into pricer

Commit:
Add spec for volume_price_model variant association

mamhoff reviewed (approved):
Much cleaner and more understandable! :)

Senjai commented on spec/models/solidus_volume_pricing/pricing_options_spec.rb:
```diff
@@ -0,0 +1,55 @@
+  end
+
+  describe 'new instances' do
+    subject(:options) do
```
Prefer keeping the subject under test known as `subject` in RSpec. Subject has a well defined meaning in tests, if we rename it, it can add confusion when we're trying to read the test as we might assume that the subject is run in a before block, and options may be some sort of side effect.

tvdeyen replied:
Sure.

Senjai commented on spec/models/solidus_volume_pricing/pricer_spec.rb:
```diff
@@ -0,0 +1,280 @@
+require 'spec_helper'
+
+RSpec.shared_examples 'having the variant price' do
```
This is fine. 

Typically though, we usually find specs one place where DRY doesn't always apply.

The benefit of using these for only one small example such as this is outweighed by the fact we have to move around the file to ascertain what's actually being run. If we had a large number of shared examples this becomes more worthwhile, but for a reader, it's probably more helpful to have it defined inline.

Senjai commented on spec/models/solidus_volume_pricing/pricer_spec.rb:
```diff
@@ -0,0 +1,280 @@
+
+RSpec.describe SolidusVolumePricing::Pricer do
+  it 'inherits from default variant pricer' do
+    expect(described_class < Spree::Variant::PriceSelector).to be(true)
```
This doesn't really provide much value as a test, but it's also completely fine.

The way we could get value from these kind of tests is by providing a shared example set to "lint" the interface of PriceSelector, as because in ruby we prefer duck typing rather that type checking.

`it_behaves_like "an implementation of a Spree::Variant::PriceSelector"` which would simply check that it responds to the public interface required by the subclass.

See how rails provides a class for verifying your "custom models" are rails compliant:

http://api.rubyonrails.org/files/activemodel/lib/active_model/lint_rb.html

tvdeyen replied:
It would be great if the core would provide this 👍

Senjai commented on spec/models/solidus_volume_pricing/pricer_spec.rb:
```diff
@@ -0,0 +1,280 @@
+      Spree::Config.volume_pricing_role = role.name
+    end
+
+    context 'discount_type = price' do
```
prefer a description that chains well when tests are ran using `--format documentation`.

Currently it would read as
```
#price_for
  discount_type = price
    when quantity does not match the range
```

Instead it would be more clear with "when the discount_type is the same as the price"

Senjai commented on spec/models/solidus_volume_pricing/pricer_spec.rb:
```diff
@@ -0,0 +1,280 @@
+        end
+      end
+
+      context 'when it does match from a volume price model' do
```
I don't super understand this statement.

tvdeyen replied:
Yeah, I was actually taking the original spec names that where already there. I will rename it into something understandable.

Senjai commented on spec/models/solidus_volume_pricing/pricer_spec.rb:
```diff
@@ -0,0 +1,280 @@
+      it_behaves_like 'having the variant price'
+    end
+
+    context 'when use_master_variant_volume_pricing' do
```
Is true? Set?

Senjai commented on spec/models/solidus_volume_pricing/pricer_spec.rb:
```diff
@@ -0,0 +1,280 @@
+      let(:master) { variant.product.master }
+
+      before do
+        @before = Spree::Config.use_master_variant_volume_pricing
```
Prefer avoiding use of instance methods.

Senjai commented on spec/models/solidus_volume_pricing/pricer_spec.rb:
```diff
@@ -0,0 +1,280 @@
+
+      context 'is enabled' do
+        before do
+          Spree::Config.use_master_variant_volume_pricing = true
```
_always_ use around here. When mutating global data in tests we have to be extremely careful to avoid any kind of leak, and we want to make it obvious we're guarding against it.

```ruby
let(:use_master_variant_volume_pricing) { true }
around do |example|
  current_pricing = Spree::Config.use_master_variant_volume_pricing
  Spree::Config.use_master_variant_volume_pricing = use_master_variant_volume_pricing
  example.run
  Spree::Config.use_master_variant_volume_pricing = current_pricing
end
```

You then override the let in your contexts as required and put this at the top level.

tvdeyen replied:
Perfectly reasonable! Will use `around` then. Thanks

Senjai commented on app/models/spree/variant_decorator.rb:
```diff
@@ -27,11 +27,6 @@ def join_volume_prices(user = nil)
   end
 
-  # calculates the price based on quantity
-  def volume_price(quantity, user = nil)
```
Will this be a major version change on this gem? Has it been released yet?

If so, can we deprecate and still return a value?

Senjai commented on app/models/spree/variant_decorator.rb:
```diff
@@ -68,17 +63,6 @@ def compute_volume_price_quantities(type, default_price, quantity, user)
     end
   end
 
-  def compute_volume_price(volume_price)
```
ditto.

Senjai commented on app/models/spree/variant_decorator.rb:
```diff
@@ -7,26 +7,6 @@
       volume_price[:amount].blank? && volume_price[:range].blank?
     }
 
-  def join_volume_prices(user = nil)
```
same question as previous commit w/r/t versions/public interface.

Senjai commented on app/models/spree/volume_price.rb:
```diff
@@ -14,6 +14,16 @@ class Spree::VolumePrice < ActiveRecord::Base
   validate :range_format
 
+  def self.for_variant(variant, user: nil)
+    role_id = user ? user.resolve_role : nil
```
simplified with `user.try!(:resolve_role)` and eliminates a temporary variable.

tvdeyen replied:
I kinda think it actually provides better readability by using a local var vs. putting an expression into a method call.

Senjai commented on app/models/spree/volume_price.rb:
```diff
@@ -14,6 +14,16 @@ class Spree::VolumePrice < ActiveRecord::Base
+      (arel_table[:variant_id].eq(variant.id)
+        .or(arel_table[:volume_price_model_id].in(variant.volume_price_models.ids)))
+        .and(arel_table[:role_id].eq(role_id))
+    ).order(position: :asc)
```
Just verifying intended behaviour here is `(variant_id || price_models.include?(model_id)) && role_id`

tvdeyen replied:
The spec will

Senjai commented on app/models/spree/volume_price.rb:
```diff
@@ -14,6 +14,16 @@ class Spree::VolumePrice < ActiveRecord::Base
 
   validate :range_format
 
+  def self.for_variant(variant, user: nil)
```
Test required

tvdeyen replied:
It's very well covered by the pricer spec and I already have 100% coverage of the `volume_price` class. I will add one if you insist, though.

tvdeyen replied:
You actually convinced me IRL ;)

Senjai commented on app/models/solidus_volume_pricing/pricer.rb:
```diff
@@ -7,18 +7,22 @@ def self.pricing_options_class
     private
 
+    def extract_options(pricing_options)
+      @quantity = pricing_options.quantity
```
This can lead to questionable behaviour if people aren't paying attention. Usually, we only set instance variables in initialization. 

Doing it elsewhere without the method name explicitly stating that it is "setting" something on the class, or for memoization (usually we indicate memoization with @_somthing to indicate we're not using the instance variable for anything else)

Moreso the method name indicates that it should return "options", either as an array, a hash, or a structure.

It's okay to not set this as instance variables, and make private methods take an argument.

Senjai commented on app/models/spree/variant_decorator.rb:
```diff
@@ -12,11 +12,6 @@ def volume_price_earning_percent(quantity, user = nil)
   end
 
-  # return amount of earning
-  def volume_price_earning_amount(quantity, user = nil)
```
same question w/r/t interface maintenance.

Senjai commented on spec/models/solidus_volume_pricing/pricer_spec.rb:
```diff
@@ -277,4 +278,203 @@
+      SolidusVolumePricing::PricingOptions.new(quantity: quantity)
+    end
+
+    subject(:earning_amount) do
```
subject naming conversation earlier.

Senjai commented on app/helpers/spree/base_helper_decorator.rb:
```diff
@@ -1,17 +1,19 @@
-      variant.volume_price_earning_amount(quantity, user),
-      currency: Spree::Config[:currency]
-    ).to_html
+    price_display(variant, quantity: quantity, user: user).earning_string
```
+1 for creating an object to handle this.

However replacing `price_display` with `SolidusVolumePricing::PriceDisplay` actually will do the same thing. They take the same arguments and the method has no logic, so we win on clarity by just explicitly referencing the class.

Senjai commented on app/helpers/spree/base_helper_decorator.rb:
```diff
@@ -4,11 +4,11 @@ def display_volume_price(variant, quantity = 1, user = nil)
 
   def display_volume_price_earning_percent(variant, quantity = 1, user = nil)
-    variant.volume_price_earning_percent(quantity, user).round.to_s
+    price_display(variant, quantity: quantity, user: user).earning_percent_string
```
same comment re price display

Senjai commented on app/models/solidus_volume_pricing/price_display.rb:
```diff
@@ -12,8 +12,12 @@ def price_string
       price.to_s
     end
 
-    def earning_string
```
Prefer fixing the original commit, as it was added in this PR

Senjai commented on spec/models/spree/variant_spec.rb:
```diff
@@ -1,3 +1,4 @@
 RSpec.describe Spree::Variant, type: :model do
   it { is_expected.to have_many(:volume_prices) }
+  it { is_expected.to have_many(:model_volume_prices) }
```
I'm fine with these assertions if people prefer them, but I contest that they don't provide any value.

Testing that a relationship is defined and exists is superfluous to testing methods that _use_ that relationship. Which makes this redundant.

tvdeyen replied:
I absolutely agree. Was just repeating the style of testing from above. Will leave it, though :)

Commit:
Use let instead of instance var in volume price spec

By using a let and not an instance var in a before each block in the volume price spec we don’t create volume prices upfront.

Commit:
Lock bundler to 1.13.7 on travis

tvdeyen commented:
@Senjai thanks for the thorough review. I really appreciate. Addressed all of the comments.

tvdeyen commented on .travis.yml:
```diff
@@ -10,3 +10,8 @@ env:
+  - rvm use @global
+  - gem uninstall bundler -x
+  - gem install bundler --version=1.13.7
+  - bundler --version
```
This is needed as Bundler 1.14 is messed up right now https://github.com/bundler/bundler/issues/5359
